File: problems/[57]插入区间.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
	vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
		if (intervals.empty()) {
			intervals.push_back(newInterval);
			return intervals;
		}
		int left = 0;
		int right = intervals.size();
		int target = newInterval[0];
		while (left < right) {
			int mid = (left + right) / 2;
			if (intervals[mid][0] >= target)
				right = mid;
			else
				left = mid + 1;
		}
		
		int i;
		int sp;
		vector<int> cur;
		if (left == 0 || intervals[left-1][1] < newInterval[0]) {
			cur = newInterval;
			sp = left;
		}
		else {
			cur = intervals[left-1];
			cur[1] = max(cur[1], newInterval[1]);
			sp = left - 1;
		}
		for (i = left; i < intervals.size(); ++i) {
			if (cur[1] >= intervals[i][0])
				cur[1] = max(cur[1], intervals[i][1]);
			else
				break;
		}
		intervals.erase(intervals.begin() + sp, intervals.begin()+i);
		intervals.insert(intervals.begin() + sp, cur);
		return intervals;
	}
};

File: solutions/[57]插入区间.py
from typing import List


class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        # 1 找到新区间插入位置
        insert_p = len(intervals)
        for i, (_, right) in enumerate(intervals):
            if newInterval[0] > right:
                insert_p = i + 1
                break
        intervals.insert(insert_p, newInterval)

        # 2 合并重叠区间
        right = newInterval[1]
        left_p = insert_p+1
        for i in range(insert_p+1, len(intervals)):
            if right < intervals[i][0]:
                left_p = i
                break
            else:
                right = max(right, intervals[i][1])
        return intervals[:insert_p] + [[newInterval[0], right]] + intervals[left_p:]
